plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/attribute-filter/use-product-filter-clear-button-manager.tsx

```diff
@@ -0,0 +1,132 @@
+/**
+ * External dependencies
+ */
+import { useDispatch, useSelect } from '@wordpress/data';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+import { BlockAttributes, createBlock } from '@wordpress/blocks';
+import { useEffect, useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { getInnerBlockByName } from '../../utils';
+
+function findClientIdByName(
+	block: BlockAttributes,
+	targetName: string
+): string | undefined {
+	if ( ! block ) {
+		return undefined;
+	}
+
+	if ( block.name === targetName ) {
+		return block.clientId;
+	}
+
+	if ( block.innerBlocks && block.innerBlocks.length > 0 ) {
+		for ( const innerBlock of block.innerBlocks ) {
+			const blockId: string | undefined = findClientIdByName(
+				innerBlock,
+				targetName
+			);
+			if ( blockId ) {
+				return blockId;
+			}
+		}
+	}
+
+	return undefined;
+}
+
+export const useProductFilterClearButtonManager = ( {
+	clientId,
+	showClearButton,
+}: {
+	clientId: string;
+	showClearButton: boolean;
+} ) => {
+	const [ previousShowClearButtonState, setPreviousShowClearButtonState ] =
+		useState< boolean >( showClearButton );
+	// @ts-expect-error @wordpress/data types are outdated.
+	const { insertBlock, removeBlock, updateBlockAttributes } =
+		useDispatch( blockEditorStore );
+
+	const { filterBlock, clearButtonBlock, clearButtonParentBlock } = useSelect(
+		( select ) => {
+			const { getBlock, getBlockParents } = select( blockEditorStore );
+			const filterBlockInstance = getBlock( clientId );
+			const clearButtonId = findClientIdByName(
+				filterBlockInstance,
+				'woocommerce/product-filter-clear-button'
+			);
+			const clearButtonBlockInstance = clearButtonId
+				? getBlock( clearButtonId )
+				: undefined;
+			const clearButtonParentBlocks = getBlockParents(
+				clearButtonId,
+				true
+			);
+			const clearButtonParentBlockInstance =
+				clearButtonParentBlocks.length
+					? getBlock( clearButtonParentBlocks[ 0 ] )
+					: null;
+
+			return {
+				filterBlock: filterBlockInstance,
+				clearButtonBlock: clearButtonBlockInstance,
+				clearButtonParentBlock: clearButtonParentBlockInstance,
+			};
+		}
+	);
+
+	function findPositionToAddTheClearButtonBlock() {
+		if ( clearButtonParentBlock ) {
+			return {
+				clearButtonBlockPosition: 1,
+				clearButtonParentBlockId: clearButtonParentBlock?.clientId,
+			};
+		}
+		const filterBlockHeader = getInnerBlockByName(
+			filterBlock,
+			'core/group'
+		);
+		const filterBlockHeading = findClientIdByName(
+			filterBlockHeader,
+			'core/heading'
+		);
+		if ( Boolean( filterBlockHeading ) ) {
+			return {
+				clearButtonBlockPosition: 1,
+				clearButtonParentBlockId: filterBlockHeader?.clientId,
+			};
+		}
+		return {
+			clearButtonBlockPosition: 0,
+			clearButtonParentBlockId: clientId,
+		};
+	}
+
+	useEffect( () => {
```

Comment:
I wonder if we can make it a shared util instead of a hook to execute on an attribute change event. Instead of subscribe to the change, we can only do the manipulation when we need it.

Reply:
Sorry I missed your message 😞 . We can convert the hook to just be a regular function, but TBH I don't see any advantage in this case since we would have to duplicate the logic on each filter to detect the attribute change and react accordingly. The current implementation already handles that since it's hooked into the React state and lifecycle.

Reply:
> I don't see any advantage in this case

I'm commenting from the performance perspective. Because using a hook means that the callback will execute anytime dependencies change. We short-circuit most of the cases so it's probably not an issue here. Although the `useSelect` is always executed, the counterpoint is: that it's cached so it's cheap to run. I don't have any benchmark to back my concern so you can skip this comment. The performance gain (if any) might not be worth the effort, too.

> since we would have to duplicate the logic on each filter to detect the attribute change and react accordingly

I don't think so. We only need to execute the function in the `onChange` callback of the inspector control. I don't see the need to detect the attribute change here.

Reply:
Oh, now I see what you mean. Yeah, performance-wise, I don’t think we would gain much, but in terms of only calling the function when the attribute changes, I agree it would be better. I refactored part of the code and moved some functions to the product-filters/utils folder, but I’ll explore your suggestion tomorrow. I think with that, we can get rid of the hook. Thank you for sharing this!

Reply:
What do you think about refactoring it later as we're wrapping up this cycle?

Reply:
I agree with you. Let's review the current implementation then, and we can refactor it later if needed.